### DISCUSSION TITLE
https://github.com/npm/cli/issues/564 - npm ci -flag that installs only missing packages or re-installs packages which version does not match package-lock.json

### DISCUSSION THREAD

**Discussion by tamusjroyce** [npm/feedback#110] [CLI] [answered]
@darcyclarke closed this feature (created before npm / feedback)

There should be a way to run npm ci -flag or other npm magic that installs only missing packages or re-installs packages which version does not match package-lock.json

**Comment by tamusjroyce** [comment]
This type of feature would not resolve corrupt files if they were present and the packages' version matches what is in package-lock.json. Would a hash of the folder need to match the version hash. If they don't match, delete and reinstall? (similar to how emerge works for gentoo - unrelated except it is also a package manager)

**Reply by tamusjroyce** [reply]
As deno evolves to support npmjs, this features is no longer necessary or needed

**Comment by MylesBorins** [comment, marked as answer]
My understanding is that the `npm ci` algorithm was designed in such a way that it is assuming that there is no `node_modules` folder. As such it seems like a flag for `npm ci` isn't really a fit (and what it seems was the conclusion in the referenced thread).

It seems to me like the ask is closer to an `npm install` that doesn't modify the package-lock... but it isn't super clear to me just yet. Can you outline the usecase you are trying to fulfill a bit more?

**Reply by MylesBorins** [reply]
If you want to avoid changes to your package.json & package-lock.json when running install run `npm install --no-save`

**Reply by tamusjroyce** [reply]
Surprised this didn't pop up as a notification. Sorry I didn't reply. 
1. `npm install` is driven off of package.json to install node modules instead of package-lock.json
2. `npm ci` is driven off of package-lock.json to install files

At least at the time, I was wanting package-lock.json to drive only missing packages in node_modules. A majority of our CI pipeline is spent running `npm ci` because it wipes node_modules clean before installing (at least for our dev CI pipelines - QA & higher probably should wipe node_modules for stability).

I'm not sure the state of the latest npm and assume it has some similar feature by now. Thank you for the response and the great work that has been done!

**Comment by nfriedly** [comment]
@MylesBorins 

Does `npm install --no-save` actually ensure the dependency tree matches `package-lock.json` or does it just do the regular `npm install` thing and then skip the final step of updating `package-lock.json`?

Right now it seems like the only way to keep a team in sync is to have everyone run `npm ci` any time a dependency changes, which wastes a lot of time because it's installing everything from scratch. I'd rather have everyone run `npm install` because it's so much faster to update, but that inevitably leads to useless changes to `package-lock.json`. No one is trying to change the dependencies, just update their environment to have the correct set of dependencies, so `package-lock.json` shouldn't be changing.

What I want is something that updates `node_modules/` to match the `package-lock.json` without wasting all of the time deleting and re-downloading all of the things that are already correct.
